### PR TITLE
Fixed Attachments Previews

### PR DESCRIPTION
**Pull request by buh**
**Internal changes**
- Fixed attachments preview
- Adopted for iOS 13
- Cleaned up the code: protocol, names.

Check screenshots in CIS-102

#skip_changelog

### CODE REVIEW

File: Sources/UI/Chat/Attachments/AttachmentPreview.swift [hunk 1]
```diff
@@ -1,353 +1,31 @@
...
 //
-//  Created by Alexey Bukhtin on 09/04/2019.
+//  Created by Alexey Bukhtin on 12/04/2019.
 //  Copyright © 2019 Stream.io Inc. All rights reserved.
 //
 
 import UIKit
 import StreamChatClient
 import StreamChatCore
 import RxSwift
```
Comment (b-onc) [inline Sources/UI/Chat/Attachments/AttachmentPreview.swift:12]: We don't need RxSwift dependency in this file

File: Sources/UI/Chat/Attachments/AttachmentPreview.swift [hunk 2]
```diff
@@ -1,353 +1,31 @@
...
-final class AttachmentPreview: UIView, AttachmentPreviewProtocol {
+protocol AttachmentPreview where Self: UIView {
+    typealias Completion = (AttachmentPreview, Error?) -> Void
     
-    private var defaultHeight: CGFloat = .attachmentPreviewMaxHeight
+    /// An index in the message stack view.
+    var index: Int { get set }
+    /// An attachment.
+    var attachment: Attachment? { get }
+    var disposeBag: DisposeBag { get }
```
Comment (b-onc) [inline Sources/UI/Chat/Attachments/AttachmentPreview.swift:21]: Why do we need disposeBag as a protocol dependency?

File: Sources/UI/Chat/Attachments/AttachmentPreviewProtocol.swift [hunk 3]
```diff
@@ -11,9 +11,10 @@ import StreamChatClient
 import StreamChatCore
 import RxSwift
 
-protocol AttachmentPreviewProtocol where Self: UIView {
+protocol AttachmentPreview where Self: UIView {
```
Comment (b-onc) [inline Sources/UI/Chat/Attachments/AttachmentPreviewProtocol.swift:14]: Isn't this the same as the file above?
`protocol AttachmentPreview where Self: UIView`

File: Sources/UI/Chat/Cells/MessageTableViewCell+Attachments.swift [hunk 4]
```diff
@@ -40,96 +38,76 @@ extension MessageTableViewCell {
...
         
+        preview.index = index
         messageStackView.insertArrangedSubview(preview, at: index)
         attachmentPreviews.append(preview)
         
         // File preview.
-        if isFileAttachment {
-            preview.update(maskImage: backgroundImageForAttachment(at: index)) { _, _ in }
+        guard let imagePreview = preview as? ImageAttachmentPreview else {
+            preview.update { _, _ in }
```
Comment (b-onc) [inline Sources/UI/Chat/Cells/MessageTableViewCell+Attachments.swift:61]: Why do we need this empty call before the actual call?

File: Sources/UI/Chat/Attachments/AttachmentPreview.swift (hunk 2 above)
Comment (buh) [reply Sources/UI/Chat/Attachments/AttachmentPreview.swift:21]: Yes, we no need it in the protocol.

File: Sources/UI/Chat/Attachments/AttachmentPreviewProtocol.swift (hunk 3 above)
Comment (buh) [reply Sources/UI/Chat/Attachments/AttachmentPreviewProtocol.swift:14]: Yes, I renamed `AttachmentPreviewProtocol` to `AttachmentPreview`.

File: Sources/UI/Chat/Cells/MessageTableViewCell+Attachments.swift (hunk 4 above)
Comment (buh) [reply Sources/UI/Chat/Cells/MessageTableViewCell+Attachments.swift:61]: The first update is for updating the state of the view.
The second is update the `imageMask`.

Comment (b-onc) [reply Sources/UI/Chat/Cells/MessageTableViewCell+Attachments.swift:61]: Hmm, this seems like a code smell. Why do we need to call `update` twice? Can we combine the logic?

Comment (buh) [reply Sources/UI/Chat/Cells/MessageTableViewCell+Attachments.swift:61]: It was a single method with several logics inside, I separated it and now it's easy to manage.

Comment (buh) [reply Sources/UI/Chat/Cells/MessageTableViewCell+Attachments.swift:61]: Do you suggest anything? I know it needs time to understand. Ask me.

Comment (b-onc) [reply Sources/UI/Chat/Cells/MessageTableViewCell+Attachments.swift:61]: `update(imageMask:)` could be renamed to `apply(imageMask:)`
And in `func update(_ completion: @escaping Completion)`, parameter `completion` could be optional, so we don't need to do `preview.update { _, _ in }`, just `preview.update()`

Comment (buh) [reply Sources/UI/Chat/Cells/MessageTableViewCell+Attachments.swift]: Done.

File: Sources/UI/Chat/Cells/MessageTableViewCell+Attachments.swift [hunk 5]
```diff
@@ -56,18 +56,18 @@ extension MessageTableViewCell {
         messageStackView.insertArrangedSubview(preview, at: index)
         attachmentPreviews.append(preview)
         
-        // File preview.
         guard let imagePreview = preview as? ImageAttachmentPreview else {
-            preview.update { _, _ in }
-            preview.update(imageMask: backgroundImageForAttachment(at: index))
+            // File preview.
+            (preview as? FileAttachmentPreview)?.update()
```
Comment (b-onc) [inline Sources/UI/Chat/Cells/MessageTableViewCell+Attachments.swift:61]: Why don't you make the parameter optional in protocol so you don't need to cast?

Comment (buh) [reply Sources/UI/Chat/Cells/MessageTableViewCell+Attachments.swift:61]: I can't provide a default value anyway. it will be: `update(nil)`

Comment (buh) [reply Sources/UI/Chat/Cells/MessageTableViewCell+Attachments.swift:61]: Do you think an optional completion block is better?

Comment (b-onc) [reply Sources/UI/Chat/Cells/MessageTableViewCell+Attachments.swift:61]:
```swift
protocol A {
    func update(_ completion: () -> ())
}

extension A {
    func update(_ completion: () -> () = { }) {
        update(completion)
    }
}

class B: A {
    func update(_ completion: () -> () = { }) {
        // ...
    }
}
```
better if it's optional

Comment (b-onc) [reply Sources/UI/Chat/Cells/MessageTableViewCell+Attachments.swift:61]: > Do you think an optional completion block is better?
Definitely for this case

**Pull request merged**